tower-ready-cache/src/cache.rs

```diff
@@ -233,24 +233,38 @@ where
                 Poll::Ready(None) => return Poll::Ready(Ok(())),
                 Poll::Ready(Some(Ok((key, svc, cancel_rx)))) => {
                     trace!("endpoint ready");
-                    let cancel_tx = self
-                        .pending_cancel_txs
-                        .swap_remove(&key)
-                        .expect("missing cancelation");
-                    // Keep track of the cancelation so that it need not be
-                    // recreated after the service is used.
-                    self.ready.insert(key, (svc, (cancel_tx, cancel_rx)));
+                    let cancel_tx = self.pending_cancel_txs.swap_remove(&key);
+                    if let Some(cancel_tx) = cancel_tx {
+                        // Keep track of the cancelation so that it need not be
+                        // recreated after the service is used.
+                        self.ready.insert(key, (svc, (cancel_tx, cancel_rx)));
+                    } else {
+                        // This should not technically be possible. We must have decided to cancel
+                        // a Service (by sending on the CancelTx), yet that same service then
+                        // returns Ready. Since polling a Pending _first_ polls the CancelRx, that
+                        // _should_ always see our CancelTx send. Yet empirically, that isn't true:
+                        //
+                        //   https://github.com/tower-rs/tower/issues/415
+                        //
+                        // So, we instead detect the endpoint as canceled at this point. That
+                        // should be fine, since the oneshot is only really there to ensure that
+                        // the Pending is polled again anyway.
+                        debug!("canceled endpoint removed when ready");
```

Comment:
Do you think a `debug_assert!()` is suitable in these cases so that we might still be able to catch this in development?

Reply:
Yeah, good thinking! Added.